src/runtime_src/core/pcie/tools/cloud-daemon/container/container.cpp

```diff
@@ -0,0 +1,235 @@
+ * Output:
+ *        none
+ * Return value:
+ *        0: success
+ *        1: failure
+ */ 
+int xclLoadXclBin(size_t index, const axlf *&xclbin)
+{
+    auto d = std::make_unique<Container>(index);
+    if (!d->isGood())
+        return 1;
+    return d->xclLoadXclBin(xclbin);
+}
+
+int Container::xclLoadXclBin(const xclBin *&buffer)
+{
+    /*
+     * This file is delivered by default not to provide xclbin protection.
+     * That means, the input xclbin file is downloaded as-is. This is also
+     * useful in xrt container deployment case, with which, only the user PF
+     * is granted to the container, and mgmt is left at host, but within
+     * container users can use xbutil and/or openCL without seeing any
+     * difference compared to within host.
+     *
+     * If container platform vendors, say Nimbix, want to have xclbin proction,
+     * their code can be added here.
+     */
+#if 1
+    xclmgmt_ioc_bitstream_axlf obj = { const_cast<axlf *>(buffer) };
+#else
+    //add vendor specific code here
+    std::shared_ptr<std::vector<char>> real_xclbin;
+    if (retrieve_xclbin(buffer, real_xclbin) != 0)
+        return -EINVAL;
+    xclmgmt_ioc_bitstream_axlf obj = {reinterpret_cast<axlf *>(real_xclbin.get()->data())};    
+#endif
+    return mgmtDev->ioctl(XCLMGMT_IOCICAPDOWNLOAD_AXLF, &obj);
+}
+
+bool Container::isGood()
+{
+    return mgmtDev != nullptr;
+}
+
+Container::~Container()
+{
+}
+
+Container::Container(size_t index)
+{
+    mgmtDev = pcidev::get_dev(index, false);
+}
+
+//private methods, vendor dependant
+/*
+ * This file also gives examples how users add customized xclbin protection
+ * leveraging the mpd/msd framework.
+ * 
+ * Entry in the xclbin repository.
+ * This is only for the sample code usage. Cloud vendor has freedom to define
+ * its own in terms of their own implementations
+ */ 
+/* The fake xclbin file transmitted through mailbox is achieved by
+ * #xclbinutil --input path_to_xclbin --remove-section BITSTREAM --output path_to_fake_xclbin
+ * --skip-uuid-insertion
+ * this new fake xclbin has same uuid to the real xclbin
+ *
+ * md5 of the fake xclbin can be achieved by
+ * #md5sum path_to_fake_xclbin
+ *
+ * This md5 is the primary key of the repo database to get the real xclbin
+ */
+struct xclbin_repo {
+    const char *md5; //md5 of the xclbin metadata. should be the primary key of DB of the repo
+    const char *path; //path to xclbin file
+};
+static struct xclbin_repo repo[2] = {
+    {
+        .md5 = "d9662fc2a45422d5f7c80f57dae4c8db",
+        .path = "/opt/xilinx/dsa/xilinx_u280_xdma_201910_1/test/verify.xclbin",
+    },
+    {
+        .md5 = "97aefd0cd3dd9a96cc5d24c9afcd3818",
+        .path = "/opt/xilinx/dsa/xilinx_u280_xdma_201910_1/test/bandwidth.xclbin",
+    },
+}; // there are only 2 xclbins in the sample code
+
+int Container::retrieve_xclbin(const xclBin *&orig,
+       std::shared_ptr<std::vector<char>> &real_xclbin)
+{
+    //go get the real_xclbin yourself
```

Comment:
Where should I go to get the `real_xclbin` myself???

Reply:
comments should have been clear.
This part is the sample code for user's reference. What this part does is from the input fake xclbin   to get the real xclbin. Regarding how to get real xclbin, it's user dependent. The user should know where to get the real xclbin themselves.